### PR TITLE
vreplication timeout query optimizer hints

### PR DESCRIPTION
**Pull request by olyazavr**
## Description
When we try to run online ddl operations using the Vitess strategy, we see certain migrations on specific keyspace/tables fail due to timeout. The [suggestion in slack](https://vitess.slack.com/archives/C0PQY0PTK/p1692205368689939) was to increase mysql's `max_execution_time`, which did the trick, but we would like to limit this to only vreplication, not affecting any other query. 

To do so, I have used the `vreplication_copy_phase_duration` flag and added `/*+ MAX_EXECUTION_TIME(x) */` where `x` is the value of `vreplication_copy_phase_duration ` to the vreplication queries. This involved moving the `vreplication_copy_phase_duration` duration flag to avoid circular dependencies

I tested this in our own environment with the previously failing schema migration and saw it make progress and complete with a 1hr setting for this flag. 

Update: This has now been running in production and we no longer see online ddl schema migrations fail because of query timeouts. We see large/busy keyspaceshards able to run migrations where they previously were unable with ghost or vitess

## Related Issue(s)
Fixes https://github.com/vitessio/vitess/issues/14081

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by olyazavr** [comment]
I got feedback to not move the flag to another package, but I get circular dependencies otherwise, not sure how to proceed here

**Comment by rohit-nayak-ps** [comment]
> I got feedback to not move the flag to another package, but I get circular dependencies otherwise, not sure how to proceed here

Sorry, meant to take a look at this earlier. My thought was to move the flag to tablet-level. I will take a look first thing tomorrow and push to this PR if that is OK. 

I can take a look at the failing unit tests too then.

**Comment by rohit-nayak-ps** [comment]
> Sorry, meant to take a look at this earlier. My thought was to move the flag to tablet-level. I will take a look first thing tomorrow and push to this PR if that is OK.

[This commit ](https://github.com/vitessio/vitess/commit/ad6b53ea5b2506d8b16272f6a47d63466f97430b) moves the flag to the vttablet level. 

> I can take a look at the failing unit tests too then.

Didn't look at the failing tests since I saw you had some recent commits to try to address them. Let me know if you want me to take a look at them.

**Comment by olyazavr** [comment]
Thanks for that commit, I've pulled it in! I took a stab at the tests, but I would love some help with the remaining ones

**Comment by rohit-nayak-ps** [comment]
> Thanks for that commit, I've pulled it in! I took a stab at the tests, but I would love some help with the remaining ones

Only one other test  related to `noblob` seems to be failing. Here is the commit that fixes it: https://github.com/vitessio/vitess/commit/99e1dfd3fd12899b96c1f0186eb3b246d72331a1.  For `noblob` there is another set of queries that need the timeout hint, which I have added in that commit.

Can you pull that and let's see if all tests pass? Thanks.

**Comment by olyazavr** [comment]
Tests are all now passing! I'm also trying to find someone with permissions to update the hubspot repo settings so others can actually push to my branches 😅

**Comment by rohit-nayak-ps** [comment]
@olyazavr, great, good to see that all tests are passing. 

Following up on the slack thread regarding this issue: did this PR fix your existing issue? Especially the fact that we are adding the timeout hints on the target for the insert/updates. Usually we have been seeing issues on the source while selecting rows in the copy phase. So just wanted to confirm that it does work.

Can you also add your specific findings to the description in the issue, just so that we have it all in one place? Thanks!

**Comment by olyazavr** [comment]
Yes, it did fix the issue, and we are now able to run migrations on larger/busier keyspaceshards that previously could not run schema migrations on ghost or vitess (: 

If the question is do we need the timeout hints for anything other than the selects - I confess I didn't try this out, and added the timeout hints in all places to be consistent, and because it seemed like there was no downside to having all vreplication queries be allowed to run longer 

I did not test this with any other usage of vreplication other than online ddl schema migrations

**Comment by rohit-nayak-ps** [comment]
@olyazavr, turns out that the `MAX_EXECUTION_TIME` hint is only applicable for `SELECT`s: https://dev.mysql.com/doc/refman/8.0/en/optimizer-hints.html#optimizer-hints-execution-time. We should revert the `tabletmanager` changes.

File: go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go [hunk 1]
```diff
@@ -726,6 +727,7 @@ func (tpb *tablePlanBuilder) generateValuesPart(buf *sqlparser.TrackedBuffer, bv
 func (tpb *tablePlanBuilder) generateSelectPart(buf *sqlparser.TrackedBuffer, bvf *bindvarFormatter) *sqlparser.ParsedQuery {
 	bvf.mode = bvAfter
 	buf.WriteString(" select ")
+	buf.WriteString(vstreamer.GetVReplicationMaxExecutionTimeQueryHint())
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go:730]: @olyazavr we don't need this as well. We are essentially selecting static values which are used in inserts. So the `tabletmanager\vreplication` package doesn't need the optimizer hints anywhere.

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

Thanks for this @olyazavr, this should be very useful to resolve issues that we have seen occassionally with users with large clusters.

**Comment by olyazavr** [comment]
Rebased and dealt with merge conflicts

**Comment by olyazavr** [comment]
Looks like all checks are passing, anything else for me to do here?

**Review by deepthi — APPROVED** [review]

**Comment by mattlord** [comment]
https://github.com/vitessio/website/pull/1797

**Pull request merged**